File: cogito/server/models/Plan.rb
class Plan
  include MongoMapper::EmbeddedDocument

  key :date,  Time,   :default => lambda { Time.new }
  key :name,  String, :defualt => ''
  key :text,  String, :required => true
  timestamps!

  def self.get(userId, planId)

    plans = self.fetch(userId)
    _plan = plans.each do |plan|
      break plan if plan._id.to_s == planId.to_s
    end

    return _plan
  end

  def self.fetch(userId)

    user = User.get(userId)
    user.plans

  end

  def self.add_plan(userId, plan)

    _plan = Plan.new(plan)
    User.push(userId, :plans => _plan.to_mongo)

    user = User.get(userId)
    
    if user.save(:safe => true)
      return user.plans
    else
      return { :errros => user.errors }
    end 

  end

  def self.update(userId, planId, update)
    plan = self.get(userId, planId)
    plan.update_attributes(update)
    if plan.save(:safe => true)
      return plan
    else
      return { :errors => plan.errors }
    end
  end
end

File: models/Plan.js
module.exports = function(config, mongoose) {

  var crypto = require('crypto');

  var PlanSchema = new mongoose.Schema({
    date: { type: Date, default: Date.now },
    name: { type: String, default: '' },
    description: { type: String, required: true }
  });

  var Plan = mongoose.model('Plan', PlanSchema);

  function add(user, plan, callback) {
    console.log('add plan');

    // user not found
    if(user === null) { callback('User not found'); };

    user.plans.push(plan);

    user.save(function(err) {
      callback(err, plan);  
    });
  };

  function get(user, planId, callback) {
    var plans = this.fetch(user),
        match = false;

    if( !planId ) {
      callback(null, plans);
      return;
    }
    else {
      plans.forEach(function(plan, index) {
        console.log(plan, planId);
        if(plan._id === planId) {
          match = true;
          callback(null, plan);
          return;
        }
      }); 
    }

    if( !match ) {
      callback('not found');
    } 
  };

  function fetch(user) {
    return user.plans;
  };

  function update(plan, update, callback) {
    Plan.update(plan, { multi: false }, function(err, numberAffected, resp) {
      callback(err, resp);
    });
  };

  return {
    _schema: PlanSchema,
    add: add,
    get: get,
    fetch: fetch,
    update: update
  };
}